pkg/openstack/nova.go

```diff
@@ -45,9 +51,110 @@ func ReconcileNova(ctx context.Context, instance *corev1beta1.OpenStackControlPl
 			return res, err
 		}
 		instance.Status.Conditions.Remove(corev1beta1.OpenStackControlPlaneNovaReadyCondition)
+		instance.Status.Conditions.Remove(corev1beta1.OpenStackControlPlaneExposeNovaReadyCondition)
 		return ctrl.Result{}, nil
 	}
 
+	// add selector to service overrides
+	for _, endpointType := range []service.Endpoint{service.EndpointPublic, service.EndpointInternal} {
+		// NovaAPI
+		if instance.Spec.Nova.Template.APIServiceTemplate.Override.Service == nil {
+			instance.Spec.Nova.Template.APIServiceTemplate.Override.Service = map[string]service.RoutedOverrideSpec{}
+		}
+		instance.Spec.Nova.Template.APIServiceTemplate.Override.Service[string(endpointType)] =
+			AddServiceComponentLabel(
+				ptr.To(instance.Spec.Nova.Template.APIServiceTemplate.Override.Service[string(endpointType)]),
+				nova.Name+"-api")
+
+		// cell NoVNCProxy service override
+		for cellName := range instance.Spec.Nova.Template.CellTemplates {
+			cellTemplate := instance.Spec.Nova.Template.CellTemplates[cellName]
+			if cellTemplate.NoVNCProxyServiceTemplate.Override.Service == nil {
+				cellTemplate.NoVNCProxyServiceTemplate.Override.Service = map[string]service.RoutedOverrideSpec{}
+			}
+			cellTemplate.NoVNCProxyServiceTemplate.Override.Service[string(endpointType)] =
+				AddServiceComponentLabel(
+					ptr.To(cellTemplate.NoVNCProxyServiceTemplate.Override.Service[string(endpointType)]),
+					getNoVNCProxyServiceLabel(nova.Name, cellName))
+			instance.Spec.Nova.Template.CellTemplates[cellName] = cellTemplate
+		}
+	}
+
+	// When component services got created check if there is the need to create a route
+	if err := helper.GetClient().Get(ctx, types.NamespacedName{Name: "nova", Namespace: instance.Namespace}, nova); err != nil {
+		if !k8s_errors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Nova API
+	if nova.Status.Conditions.IsTrue(novav1.NovaAPIReadyCondition) {
+		svcs, err := service.GetServicesListWithLabel(
+			ctx,
+			helper,
+			instance.Namespace,
+			map[string]string{common.AppSelector: nova.Name + "-api"},
+		)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		var ctrlResult reconcile.Result
+		instance.Spec.Nova.Template.APIServiceTemplate.Override.Service, ctrlResult, err = EnsureRoute(
+			ctx,
+			instance,
+			helper,
+			nova,
+			svcs,
+			instance.Spec.Nova.Template.APIServiceTemplate.Override.Service,
+			instance.Spec.Nova.APIOverride.Route,
+			corev1beta1.OpenStackControlPlaneExposeNovaReadyCondition,
+		)
+		if err != nil {
+			return ctrlResult, err
+		} else if (ctrlResult != ctrl.Result{}) {
+			return ctrlResult, nil
+		}
+	}
+
+	if nova.Status.Conditions.IsTrue(novav1.NovaAllCellsDBReadyCondition) {
```

Comment:
NovaAllCellsDBReadyCondition will be true before NovaNoVNCProxy service is created. 

The NovaAllCellReady would be the ideal target. But if the NovaCell wants to avoid updating the compute config after the NoVNCProyx EndpointURL is set[1], then NovaAllCellReady will not be true until we have the route.

So we have multiple options:
* you can look up the NovaNoVNCProxy CR and look at its ExposeServiceReadyCondition (or whathever we rename that to) status.
* the Nova CR grows a new status condition that shows when all NoVNCProxy service in all cells is ready

[1] https://github.com/openstack-k8s-operators/nova-operator/pull/522#discussion_r1324441019

Reply:
for now I switched to use NovaAllCellReady. I'd like to wait for do a change from one of the options until we are further into the TLS topic where we also update the service operator CR with certs to be used for TLS-E.

Reply:
OK. The current nova-operator impl makes the Cell ready without waiting for the EndpointURL set so the current code works. And we can refine it later.